src/transport/http_low_level.rs

```diff
@@ -0,0 +1,190 @@
+use std::pin::Pin;
+
+use async_trait::async_trait;
+
+#[cfg(feature = "hyper")]
+pub(crate) type HyperClient<C> = hyper::Client<C, hyper::Body>;
+
+#[cfg(feature = "tokio_rt")]
+pub(crate) type HttpsConnector = hyper_tls::HttpsConnector<hyper::client::HttpConnector>;
+
+#[cfg(feature = "tokio_rt")]
+pub(crate) type HyperHttpsClient = HyperClient<HttpsConnector>;
+
+#[cfg(feature = "wasm")]
+pub(crate) struct WasmClient;
+
+/// A low level HTTP trait.
+#[async_trait(?Send)]
+pub trait HttpLowLevel {
+    /// HTTP Method type.
+    type Method;
+
+    /// HTTP Header map type.
+    type HeaderMap;
+
+    /// HTTP Body type.
+    type Body;
+
+    /// HTTP Request type.
+    type Request;
+
+    /// HTTP Response type.
+    type Response;
+
+    /// Return a request which is accepted by the client.
+    fn make_request(
+        method: Self::Method,
+        uri: &str,
+        headers: Self::HeaderMap,
+        body: Self::Body,
+    ) -> Self::Request;
+
+    /// Send a request and receive a response.
+    async fn send(&self, request: Self::Request) -> Self::Response;
+}
+
+#[cfg(feature = "hyper")]
+#[async_trait(?Send)]
+impl<C> HttpLowLevel for HyperClient<C>
+where
+    C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
+{
+    type Method = http::Method;
+    type HeaderMap = http::HeaderMap;
+    type Body = Vec<u8>;
+    type Request = http::Request<hyper::Body>;
+    type Response = http::Response<hyper::Body>;
+
+    fn make_request(
+        method: Self::Method,
+        uri: &str,
+        headers: Self::HeaderMap,
+        body: Self::Body,
+    ) -> Self::Request {
+        let mut builder = http::Request::builder().method(method).uri(uri);
+        if let Some(h) = builder.headers_mut() {
+            *h = headers;
+        }
+
+        builder.body(body.into()).unwrap()
+    }
+
+    async fn send(&self, request: Self::Request) -> Self::Response {
```

Comment:
Now, that I see this, it makes sense to me to wrap the output of send in a typersense::Result, what do you think? @emmanuelantony2000

Reply:
I have not yet implemented the error handling, but yeah will wrap them in Result.